(#98) fixed 404 links to apidocs (#99)
sakuli/docs: content/writing_tests/webtests.md

# Web tests

For DOM based testing most of the functions from
<a href="https://sahipro.com/docs/sahi-apis/" target="_blank">Sahi tests</a> can be used (please note that Sakuli only
implements the open source APIs).

The main difference between Sakuli v1 and Sakuli v2 is the usage of
<a href="https://developers.google.com/web/fundamentals/primers/promises" target="_blank">Promises</a> in the action
API, meaning that you have to `await` a click for example.

On the other hand, element selectors remain synchronized functions but will not do the actual DOM fetching anymore.
While an expression like `var $e=_link('Sakuli')` did an actual DOM-access in Sakuli v1.x, it returns a kind of abstract
query for an element now. So, action can fetch this element whenever it is required.

> A detailed list of all available functions can be found in the [Legacy API interface](/apidoc/sakuli-legacy/interfaces/legacyapi.html),

Reviewer: ```suggestion
A detailed list of all available functions can be found in the [Legacy API interface](https://sakuli.io/apidoc/sakuli-legacy),
```

| `string`                                                     | The string is normalized and wrapped into a RegExp, therefore the same logic as for RegExp is applied |

> Since we mostly apply the logic of Sahi comparisons against the class attribute are pretty dumb. While the attribute value is semantically a space separated list of class names. It is just handled as a usual string in Sahi (and therefore also in Sakuli so far).

## Actions

Actions usually invoke a <a href="https://seleniumhq.github.io/selenium/docs/api/javascript/module/selenium-webdriver/lib/input_exports_Actions.html" target="_blank">Selenium action sequence</a>
with an activated bridge mode to cover compatibility to most webdriver implementations. An action accepts a
[ElementQuery or a WebElement](#ElementQueries) and tries to perform the action on this element several times.
This approach reduces the count of StaleElementReferenceErrors dramatically, especially when a query is used.

### _eval

> Beside the fact that actions work asynchronously now, they behave like in Sahi. One exception is the [`_eval`](/apidoc/sakuli-legacy/interfaces/legacyapi.html#_eval) method, which accepts a string now containing some JavaScript code, which is performed on the website by the webdriver implementation (see `executeAsyncScript` method of 

Reviewer: ```suggestion
Beside the fact that actions work asynchronously now, they behave like in Sahi. One exception is the [`_eval`](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_eval) method, which accepts a string now containing some JavaScript code, which is performed on the website by the webdriver implementation (see `executeAsyncScript` method of 
```

## Relations
List of relations:
- [_parentNode](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_parentnode)
- [_parentCell](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_parentcell)
- [_parentRow](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_parentrow)
- [_parentTable](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_parenttable)
- [_in](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_in)
- [_near](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_near)
- [_rightOf](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_rightof)
- [_leftOf](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_leftof)
- [_leftOrRightOf](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_leftorrightof)
- [_under](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_under)
- [_above](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_above)
- [_underOrAbove](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_underorabove)

## Assertions
List of assertion functions:

> - [__assert](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assert)
> - [__assertTrue](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_asserttrue)
> - [__assertFalse](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertfalse)
> - [__assertNotTrue](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnottrue)
> - [__assertContainsText](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertcontainstext)
> - [__assertNotContainsText](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnotcontainstext)
> - [__assertEqual](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertequal)
> - [__assertNotEqual](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnotequal)
> - [__assertEqualArrays](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertequalarrays)
> - [__assertExists](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertexists)
> - [__assertNotExists](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnotexists)
> - [__assertNotNull](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnotnull)
> - [__assertNull](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnull)

Reviewer: ```suggestion
- [_assert](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assert)
- [_assertTrue](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_asserttrue)
- [_assertFalse](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertfalse)
- [_assertNotTrue](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnottrue)
- [_assertContainsText](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertcontainstext)
- [_assertNotContainsText](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnotcontainstext)
- [_assertEqual](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertequal)
- [_assertNotEqual](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnotequal)
- [_assertEqualArrays](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertequalarrays)
- [_assertExists](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertexists)
- [_assertNotExists](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnotexists)
- [_assertNotNull](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnotnull)
- [_assertNull](https://sakuli.io/apidoc/sakuli-legacy/interfaces/legacyapi.html#_assertnull)
```
